liesel/goose/optim.py

```diff
@@ -0,0 +1,233 @@
+    true value of 1.
+
+    >>> coef = lsl.param(jnp.zeros(2), name="coef")
+    >>> xvar = lsl.obs(jnp.c_[jnp.ones_like(x), x], name="x")
+    >>> mu = lsl.Var(lsl.Calc(jnp.dot, xvar, coef), name="mu")
+    >>> ydist = lsl.Dist(tfd.Normal, loc=mu, scale=1.0)
+    >>> yvar = lsl.obs(y, ydist, name="y")
+    >>> model = lsl.GraphBuilder().add(yvar).build_model()
+
+    Now, we are ready to run the optimization.
+
+    >>> result = gs.optim_flat(model, params=["coef"])
+    >>> result.position
+    {'coef': Array([0.5227438, 1.2980561], dtype=float32)}
+
+    We can now, for example, use ``result.model_state`` in
+    :meth:`.EngineBuilder.set_initial_values` to implement a "warm start" of MCMC
+    sampling.
+    """
+    if optimizer is None:
+        optimizer = optax.adam(learning_rate=1e-1)
+
+    n = _find_sample_size(model)
+    observed = _find_observed(model)
+    interface = gs.LieselInterface(model)
+    position = interface.extract_position(params, model.state)
+    interface._model.auto_update = False
+
+    def _batched_neg_log_prob(position: Position, batch_indices: Array | None = None):
+        if batch_indices is not None:
+            batched_observed = batched_nodes(observed, batch_indices)
+            position = position | batched_observed  # type: ignore
+
+        updated_state = interface.update_state(position, model.state)
+        return -updated_state["_model_log_prob"].value
+
+    neg_log_prob_grad = jax.grad(_batched_neg_log_prob, argnums=0)
+
+    opt_state = optimizer.init(position)
+
+    nlp_start = _batched_neg_log_prob(position, batch_indices=None)
+    abs_change: Array | float = 100.0
+    rel_change: Array | float = 100.0
+    i = 0
+
+    if batch_size is None:
+        batch_size = n
+
+    n_batches_inside = n // batch_size
+    batch_indices = jnp.arange(0, batch_size)
+    last_batch_size = n % batch_size
+    last_batch_start = n_batches_inside * batch_size
+    last_batch_end = last_batch_start + last_batch_size
+    last_batch_indices = jnp.arange(last_batch_start, last_batch_end)
+
+    def _batched_update(i, val):
+        pos, opt_state = val
+        batch_indices_here = batch_indices + i * batch_size
+
+        grad = neg_log_prob_grad(pos, batch_indices=batch_indices_here)
+        updates, opt_state = optimizer.update(grad, opt_state)
+        pos = optax.apply_updates(pos, updates)
+        return pos, opt_state
+
+    def _last_batched_update(i, val):
+        pos, opt_state = val
+        grad = neg_log_prob_grad(pos, batch_indices=last_batch_indices)
+        updates, opt_state = optimizer.update(grad, opt_state)
+        pos = optax.apply_updates(pos, updates)
+        return pos, opt_state
+
+    def _update_cond(val):
+        i, _, _, nlp_old, nlp_new = val
+
+        abs_change = jnp.abs(nlp_old - nlp_new)
+        rel_change = jnp.abs(abs_change / nlp_old)
+
+        return (abs_change > atol) & (rel_change > rtol) & (i < maxiter)
```

Comment:
Here we are checking the absolute change and the relative change between epoch $t - 1$ and $t$. I think it would be better to introduce a patience of more epochs. It might be that in some cases we are not monotonically decreasing

Reply:
That's a very good point. So far, I did not implement patience, because the algorithm runs without creating a history. But probably it should do that, and then it would be possible to implement some patience.